labrador/src/poseidon.rs

```diff
@@ -0,0 +1,189 @@
+    partial_rounds: usize, // Number of partial rounds (S-box is applied to first element of the state)
+    alpha: u64,            // Exponent for the S-box
+    mds: Vec<Vec<Zq>>,     // MDS matrix
+    ark: Vec<Vec<Zq>>,     // Round constants
+}
+
+impl PoseidonSponge {
+    // Initialize Poseidon sponge parameters
+    pub fn new(
+        rate: usize,
+        capacity: usize,
+        full_rounds: usize,
+        partial_rounds: usize,
+        alpha: u64,
+        mds: Vec<Vec<Zq>>,
+        ark: Vec<Vec<Zq>>,
+    ) -> Self {
+        // State starts with a vector of zeros
+        let state = vec![Zq::ZERO; rate + capacity];
+        Self {
+            state,
+            rate,
+            capacity,
+            full_rounds,
+            partial_rounds,
+            alpha,
+            mds,
+            ark,
+        }
+    }
+
+    // Apply the round constants (ARK) (against algebraic attacks)
+    fn apply_ark(&self, state: &mut [Zq], round_num: usize) {
+        for (i, elem) in state.iter_mut().enumerate() {
+            // adds diffusion (non-linearity)
+            *elem += self.ark[round_num][i];
+        }
+    }
+
+    // Apply the S-box (x -> x^alpha)
+    fn apply_s_box(&self, state: &mut [Zq], is_full_round: bool) {
+        if is_full_round {
+            for elem in state {
+                // Apply transformation to each element of state
+                *elem = elem.pow(self.alpha);
+            }
+        } else {
+            // Apply transformation only to the first element
+            state[0] = state[0].pow(self.alpha);
+        }
+    }
+
+    // Apply the MDS (Maximum Distance Separable) matrix
+    fn apply_mds(&self, state: &mut [Zq]) {
+        let mut new_state = state.to_vec();
+
+        // Matrix multiplication with the state for diffusion
+        for (i, _cur) in state.iter().enumerate() {
+            let mut sum = Zq::ZERO;
+            for (j, elem) in state.iter().enumerate() {
+                // matrix multiplication
+                sum += *elem * self.mds[i][j];
+            }
+            new_state[i] = sum;
+        }
+
+        state.copy_from_slice(&new_state);
+    }
+
+    // Apply a permutation
+    fn permute(&mut self) {
+        let full_rounds_half = self.full_rounds / 2;
+        let mut state = self.state.clone();
+        // full round ark/S-box/mds
+        for i in 0..full_rounds_half {
+            self.apply_ark(&mut state, i);
+            self.apply_s_box(&mut state, true);
+            self.apply_mds(&mut state);
+        }
+        // partial round ark/S-box/mds
+        for i in full_rounds_half..(full_rounds_half + self.partial_rounds) {
+            self.apply_ark(&mut state, i);
+            self.apply_s_box(&mut state, false); // Apply the S-box to the first element only
+            self.apply_mds(&mut state);
+        }
+        // full round ark/S-box/mds
+        for i in (full_rounds_half + self.partial_rounds)..(self.full_rounds) {
```

Comment:
```suggestion
        for i in (full_rounds_half + self.partial_rounds)..(self.partial_rounds + self.full_rounds) {
```
I think it should be like that. Please double check

Reply:
self.full_rounds already encompasses all the rounds, so I think adding the partial rounds would result in more rounds than it should.